### ISSUE TITLE
dabbrev-expand broken

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#143]
A Bug in Leo editor 5.0 about dabbrev-expand
https://groups.google.com/d/msg/leo-editor/5yVJzLzeOtA/atuE-qhJll8J

On Tue, Dec 30, 2014 at 8:33 PM, 翅膀 <wzy3721@gmail.com> wrote:

    *
    editor panel text

    main
    m
     
    *  press  alt + /

    minibuffer show :

    dabbrev-expand: main

    * press enter

            editor panel text
           
            main
            m

**Comment by tbnorth** [comment]
Not a problem, but personally I prefer the QCompleter UI for
completion.  In fact I use this code in a @script to monkey patch
dabbrev:
```python
def dynamicExpandHelper(event,prefix,aList,w,c=c):
    
    w = c.frame.body.wrapper.widget
    qcompleter = w.init_completer(aList)

c.abbrevCommands.dynamicExpandHelper = dynamicExpandHelper
```
which seems to be working. Just looking at it without remembering the code, it's hard to see how it does anything at all :-)

I think it would be good and probably a major code cleanup if Leo supported tab-key/completion-tab or QCompleter completion, selectable with a setting. I.e. the selection display code should be separated from the list / choice generation code well enough that implementing a completer interface in Leo is trivial.  Obviously generating lists of completions isn't always trivial, but allowing the user to navigate them and select from them should be.

**Comment by edreamleo** [comment]
On Fri, Mar 6, 2015 at 8:39 AM, tbnorth <notifications@github.com> wrote:

> Not a problem, but personally I prefer the QCompleter UI for
> completion. In fact I use this code in a @script
> <https://github.com/script> to monkey patch
> dabbrev:
>
> def dynamicExpandHelper(event,prefix,aList,w,c=c):
>
>     w = c.frame.body.wrapper.widget
>     qcompleter = w.init_completer(aList)
>
> c.abbrevCommands.dynamicExpandHelper = dynamicExpandHelper
>
> which seems to be working. Just looking at it without remembering the
> code, it's hard to see how it does anything at all :-)
>
> I think it would be good and probably a major code cleanup if Leo
> supported tab-key/completion-tab or QCompleter completion, selectable with
> a setting.
>
​All patches gratefully accepted, especially from you, Terry.

EKR

**Comment by edreamleo** [comment]
Fixed at rev 2f474c6.  Added two unit tests.

Changed code of both ``Alt-/ (dabbrev-expands)`` and ``Alt-Ctrl-/ (dabbrev-completion)``. The new code works as expected. Please reopen if further problems appear.
